packages/datadog-plugin-couchbase/src/index.js

```diff
@@ -0,0 +1,138 @@
+'use strict'
+
+const Tags = require('../../../ext/tags')
+const Kinds = require('../../../ext/kinds')
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+
+function startQuerySpan (tracer, config, query) {
+  const childOf = tracer.scope().active()
+  const span = tracer.startSpan('couchbase.query', {
+    childOf,
+    tags: {
+      'db.type': 'couchbase',
+      'span.type': 'sql',
+      'component': 'couchbase',
+      'service.name': config.service || `${tracer._service}-couchbase`,
+      'resource.name': query,
+      [Tags.SPAN_KIND]: Kinds.CLIENT
+    }
+  })
+
+  analyticsSampler.sample(span, config.analytics)
+
+  return span
+}
+
+function onRequestFinish (emitter, span) {
+  let rowsListener // eslint-disable-line prefer-const
+
+  const errorListener = (err) => {
+    span.setTag(Tags.ERROR, err)
+    span.finish()
+    emitter.removeListener('rows', rowsListener)
+  }
+
+  rowsListener = () => {
+    span.finish()
+    emitter.removeListener('error', errorListener)
+  }
+
+  emitter.once('rows', rowsListener)
+  emitter.once('error', errorListener)
+}
+
+function createWrapMaybeInvoke (tracer) {
+  return function wrapMaybeInvoke (_maybeInvoke) {
+    return function maybeInvokeWithTrace (fn, args) {
+      const scope = tracer.scope()
+      const callbackIndex = args.length - 1
+      const callback = args[callbackIndex]
+
+      if (callback instanceof Function) {
+        args[callbackIndex] = scope.bind(callback)
+      }
+
+      return scope.bind(_maybeInvoke).call(this, fn, args)
+    }
+  }
+}
+
+function createWrapQuery (tracer) {
+  return function wrapQuery (query) {
+    return function queryWithTrace (q, params, callback) {
+      const scope = tracer.scope()
+
+      if (params instanceof Function) {
+        params = scope.bind(params)
+      } else if (callback) {
+        callback = scope.bind(callback)
```

Comment:
This was a catch-all since in Couchbase v2.4.x - 2.5.x, `_cbas` calls `_cbasReq` without `maybeInvoke`. Also, the request (I assume you mean `n1qlReq`) is currently not wrapping the callback because it doesn't have access to it.

I can get rid of this `query` wrapper and write a specific wrapper for `_cbasReq` for versions `[2.4.x  - 2.5.x]`.